lib/streamlit/connections/snowflake_connection.py

```diff
@@ -0,0 +1,217 @@
+        def _query(sql: str) -> pd.DataFrame:
+            cur = self._instance.cursor()
+            cur.execute(sql)
+            return cur.fetch_pandas_all()
+
+        return _query(sql)
+
+    def write_pandas(
+        self,
+        df: pd.DataFrame,
+        table_name: str,
+        database: str | None = None,
+        schema: str | None = None,
+        chunk_size: int | None = None,
+        **kwargs,
+    ) -> tuple[  # This return type is somewhat horrifying but is the return type of pandas_tools.write_pandas, so we keep it consistent here.
+        bool,
+        int,
+        int,
+        Sequence[
+            tuple[
+                str,
+                str,
+                int,
+                int,
+                int,
+                int,
+                str | None,
+                int | None,
+                int | None,
+                str | None,
+            ]
+        ],
+    ]:
+        """Call snowflake.connector.pandas_tools.write_pandas with this connection.
+
+        This convenience method is simply a thin wrapper around the ``write_pandas``
+        function of the same name from ``snowflake.connector.pandas_tools``. For more
+        information, see the `Snowflake Python Connector documentation
+        <https://docs.snowflake.com/en/developer-guide/python-connector/python-connector-api#write_pandas>`_.
+        """
+        from snowflake.connector.pandas_tools import write_pandas  # type: ignore
+
+        return write_pandas(  # type: ignore
+            conn=self._instance,
+            df=df,
+            table_name=table_name,
+            database=database,
+            schema=schema,
+            chunk_size=chunk_size,
+            **kwargs,
+        )
+
+    def cursor(self) -> "SnowflakeCursor":
+        """Return a PEP 249-compliant cursor object.
+
+        For more information, see the `Snowflake Python Connector documentation
+        <https://docs.snowflake.com/en/developer-guide/python-connector/python-connector-api#object-cursor>`_.
+        """
+        return self._instance.cursor()
+
+    @property
+    def raw_connection(self) -> "InternalSnowflakeConnection":
+        """Access the underlying Snowflake Python connector object.
+
+        Information on how to use the Snowflake Python Connector can be found in the
+        `Snowflake Python Connector documentation<https://docs.snowflake.com/en/developer-guide/python-connector/python-connector-example>`_.
+        """
+        return self._instance
+
+    def session(self) -> "Session":
+        """Create a new Snowpark Session from this connection.
+
+        Information on how to use Snowpark sessions can be found in the `Snowpark documentation
+        <https://docs.snowflake.com/en/developer-guide/snowpark/python/working-with-dataframes>`_.
+        """
+        from snowflake.snowpark.session import Session
+
+        return cast(
+            Session, Session.builder.configs({"connection": self._instance}).create()
+        )
```

Comment:
And we believe that this version will be thread safe, right?

Reply:
Fairly sure -- the `Session` object itself not being thread safe should be circumvented by creating a new session object each time this method is called

There is some potential for a user to save a Snowpark session in session state or something like that and run into issues, but the pattern would be quite unnatural if you're using `st.connection` already